### ISSUE TITLE
Using Vizier in distributed setting

### ISSUE THREAD

**Issue by niemiaszek** [google/vizier#1215]
Hi. I'm still pretty new to Vizier, yet I'm very hopeful to use it for my optimization problem.

I found it pretty interesting that there are little to no examples of how to use Vizier OSS in the common setups. The whole service looks amazing, but it can be pretty troubling to setup. It can be seen in the recently posted issues.

I understand that the service allows for an amazing configurability. However I think that it wouldn't hurt if docs had some examples for common scenarios.

## Distributed Scenario
To begin, I'm setting up optimization for function using one CPU core for evaluation. I have a few nodes accessible, so I decided to use `ray` for distributing work.  Also, some of my suggestions are infeasible, due to the fact that conditional search spaces are not supported by any algorithm, as in #1201. In short, my each cpu-core worker has it's own Vizier Client and yields a single suggestion at the time. When the suggestion is infeasible, worker just requests another suggestion. I've setup it in this way to hopefully maximize resource usage and easily scale for my few nodes.

I have few problems with this setup. I'm not sure at all if this is the recommended way to setup such service for my case (multi-node evaluation). Only thing I found in the docs was recommendation of ray jobs for distributed workloads.

Furthermore, I setup it in such a way, avoiding batch suggestions, so I could comply to all methods (like multi-objective `gp-bandit`). However, methods for single and batch suggestions differ greatly eg. for `gp-ucb-pe`.

To summarize:
I'm wondering if there are any suggested patterns for such setups as mine (using batch vs single suggestions, setting up distributed workload with one client for core).

**Comment by xingyousong** [comment]
Hi @niemiaszek, in-line responses below:

`little to no examples of how to use Vizier OSS in the common setups` 
 * Are [these docs](https://oss-vizier.readthedocs.io/en/latest/guides/index.html#for-users) insufficient?

`Also, some of my suggestions are infeasible, due to the fact that conditional search spaces are not supported by any algorithm`
  * Are you creating an actual conditional search space? If so, I wouldn't recommend this due to lack of support. If you really do need a conditional space, you can for now define the equivalent flat search space by explicitly allowing all parameters.

`When the suggestion is infeasible, worker just requests another suggestion`
  * Generally, you will need to mark it as infeasible too, via `trial_client.complete(vz.Measurement(), infeasibility_reason='Outside of range.')` , see [here](https://oss-vizier.readthedocs.io/en/latest/guides/user/search_spaces.html#infeasibility). 

`I setup it in such a way, avoiding batch suggestions, so I could comply to all methods (like multi-objective gp-bandit). However, methods for single and batch suggestions differ greatly eg. for gp-ucb-pe.`
  * Any reason for doing so? If your function is multi-objective and you're OK with sequential optimization, just use GP-Bandit. If it's single-objective with batched optimization, just use GP-UCB-PE. Right now there's no GP-method which supports both batched and multi-objective settings. But there's also no reason to have your setting support all algorithms under the sun.

**Comment by niemiaszek** [comment]
Hi @xingyousong, thanks for responses. Sorry for attaching related code only to the second issue. I wanted to clarify:
**Are [these docs](https://oss-vizier.readthedocs.io/en/latest/guides/index.html#for-users) insufficient?**
* These docs are great as a starting point. However, I couldn't find more detailed end-to-end example in docs for distributed case, actually running some usual setup and maybe showing some best practices. I've stumbled upon [AutoML 2023 tutorial](https://oss-vizier.readthedocs.io/en/latest/tutorials/automl_conf_2023.html) and it's pretty informative, if only there was section on designing distributed optimization setup.
* Some ambiguities in the docs, quite easy to solve for the user:
1. Instead of `infeasibility_reason` from [docs](https://oss-vizier.readthedocs.io/en/latest/guides/user/search_spaces.html#infeasibility) and [source](https://github.com/google/vizier/blob/main/vizier/_src/pyvizier/shared/trial.py#L593) I had to use `infeasible_reason` on client's suggestion, as in [clients source](https://github.com/google/vizier/blob/main/vizier/_src/service/clients.py#L68). 
2. Works only with local setup (as endpoint ends up by default being `localhost:port`). Connection timeout error from `grcp` was quite generic.
```python
server = servers.DefaultVizierServer()  # Ideally created on a separate process such as a server machine.
clients.environment_variables.server_endpoint = server.endpoint  # Server address.
another_study_client = clients.Study.from_resource_name(study_client.resource_name)

```
3. This will yield same trial across many clients used in parallel, if `client_id` argument is not used for `suggest()` method. I think it's specified in the paper, but not so obvious.
```
another_study_client = clients.Study.from_resource_name(study_client.resource_name)
suggestions = study_client.suggest(count=1)
```

**Are you creating an actual conditional search space?**
* I'm using flat search space as with your suggestion. All I mean is that I'm taking infeasible trials into the account.

**Generally, you will need to mark it as infeasible too**
* I'm doing exactly that and continuing with next suggestion on the same worker. My search space is flat, but many trials can be infeasible and quickly discarded. 

**Any reason for doing so?**
* Yeah, that's currently the main point of my issue: 
1. I would want to maximize CPUs usage. If I use single suggestion, I can easily discard it if infeasible and get another one for the worker. The same worker will report result and draw another suggestion. This is contrary to getting batch of suggestions, waiting for all trials to complete (with some CPU cores just idling if drawn infeasible) and then getting another batch of suggestions. This assumes I get infeasible configurations quite often, but you use _infeasibility warping to prevent the optimizer from re-exploring infeasible regions_, so maybe I shouldn't worry that much.
2. I would like to leave a chance in the same distributed setup for swapping between `gp-ucb-pe` for single objective and `gp-bandit` for multi-objective, because I'm basically optimizing for method's maximal `accuracy` **or** for `accuracy/delay` trade-off. I'd like to test both approaches if possible and single trial API is more flexible with my parallel workers strategy. 
* My question is if I will notice slower convergence when using ex. `gp-ucb-pe` with 256 parallel workers, each obtaining single suggestion opposed to getting one batched suggestion of 256 and then distributing it for specific workers.
* From your last comment, I understand that it makes not much sense to use `gp-bandit` for such parallel setup. Maybe just quasi-random search will work for Multi Objective. I also understand that it's the best to use batch suggestions for `gp-ucb-pe` in such setup, to let _pure exploration_ do it's work. I will test most of these things myself, but it would be nice to have short description of the [supported algorithms](https://oss-vizier.readthedocs.io/en/latest/guides/user/supported_algorithms.html) and their intended use scenario. Users could have easier time to make "educated guess" on how to setup vizier and what which algorithm might work best for their case.

Sorry for long follow-up. Thanks for your help!

**Comment by xingyousong** [comment]
Understood, some action items we've taken:

1. Updated the [documentation on distributed tuning](https://oss-vizier.readthedocs.io/en/latest/guides/user/distributed.html) and explicit `client_id`.
2. [GP-UCB-PE will soon support multiobjective optimization](https://github.com/google/vizier/pull/1217) and we'll push this to the next version (0.1.21) of Vizier once ready. This should make GP-Bandit obsolete too, and avoids the need for weird "algorithm switching" tricks.

Closing this for now.

**Comment by niemiaszek** [comment]
Amazing! Thank you!
